device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/fileupload/FileUpload.java

```diff
@@ -130,7 +130,7 @@ public synchronized void uploadToBlobAsync(
         newUpload.setTask(taskScheduler.submit(fileUploadTask));
     }
 
-    private final class FileUploadStatusCallBack implements IotHubEventCallback
+    private static final class FileUploadStatusCallBack implements IotHubEventCallback
```

Comment:
this class is mocked here and below: https://github.com/Azure/azure-iot-sdk-java/blob/master/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/fileupload/FileUploadTest.java#L361
which now fails (since it is now a private static inner class, and mockit is unable to mock it): 
"java.lang.IllegalArgumentException: FileUploadStatusCallBack is not an inner class"

Since these tests rely on performing an action on the mocked callback, I don't see a way to refactor it without mocking the said callback.

Suggestions are welcome.

Reply:
It may be easier to just take this inner class and put it in it's own file. If it is in its own class file, it wouldn't have to be static. That should make it easier to mock, right?

Reply:
That makes sense, sure. Let me give that a shot 👍🏽

Reply:
Without knowing too much about this specific test, I see our choices as:

1. Make the base class accessible to the test (or the world, if necessary).
2. Drop the test.
3. Implement what was in the base class in the test class.

Which of those aren't actual options? What other options do you see?

Reply:
The inner class under question here is an implementation of our public IoT hub event callback interface.
1. I don't really see a value in making the class public, just because it is only ever invoked by our sdk internally, and users wouldn't gain anything from having access to it. Updating the inner class to not be enclosed within a parent class is an option.
Having a public interface and a private implementation is useful when mocking classes for unit tests; however, with this being a private inner class makes it complicated (plus mockit support). Looking up online, it looks like powermock (another mocking framework) does have support for this feature though.
2. Last resort.
3. I am not sure what this means - implement the sourcecode in the test class?

Reply:
For 3, this inner class is actually used in our production code, so it can't just be moved to test code

Reply:
So moving to its own class isn't that straightforward, since these classes rely on shared private variables

Reply:
I don't mean moved to test code, just copied. That said, it sounds like Tim's suggestion is best anyway.